File: music-infrastructure/src/main/java/com/music/sale/persistence/product/ProductCommandPersistenceAdapter.java
package com.music.sale.persistence.product;

import com.music.sale.domain.product.ProductItem;
import com.music.sale.persistence.product.entity.ProductItemEntity;
import com.music.sale.persistence.product.mapper.ProductCommandPersistenceMapper;
import com.music.sale.persistence.product.repository.ProductItemCommandRepository;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Repository
@Transactional
public class ProductCommandPersistenceAdapter {

    private final ProductItemCommandRepository commandRepo;
    private final ProductCommandPersistenceMapper mapper;

    public ProductCommandPersistenceAdapter(ProductItemCommandRepository commandRepo,
                                            ProductCommandPersistenceMapper mapper) {
        this.commandRepo = commandRepo;
        this.mapper = mapper;
    }

    public ProductItem save(ProductItem item) {
        ProductItemEntity entity = mapper.toEntity(item);
        ProductItemEntity saved = commandRepo.save(entity);
        return mapper.toDomain(saved);
    }

    public ProductItem update(Long id, ProductItem item) {
        // 전체 업데이트(merge) — 존재 확인
        ProductItemEntity current = commandRepo.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("ProductItem not found: " + id));

        // 덮어쓰기: id 고정
        ProductItemEntity patch = mapper.toEntity(item);
        patch.setId(current.getId());

        ProductItemEntity saved = commandRepo.save(patch);
        return mapper.toDomain(saved);
    }

    public void deleteById(Long id) {
        commandRepo.deleteById(id);
    }
}

File: music-infrastructure/src/main/kotlin/com/music/sale/persistence/product/ProductCommandPersistenceAdapter.kt
package com.music.sale.persistence.product

import com.music.sale.application.product.dto.SaveProductItemCondition
import com.music.sale.application.product.port.outport.ProductCommandPort
import com.music.sale.domain.product.Product
import com.music.sale.persistence.product.mapper.ProductCommandPersistenceMapper
import com.music.sale.persistence.product.repository.ProductCatalogCommandJpaRepository
import com.music.sale.persistence.product.repository.ProductCatalogQueryJooqRepository
import com.music.sale.persistence.product.repository.ProductItemCommandJpaRepository
import org.springframework.stereotype.Component
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
@Component
@Transactional
open class ProductCommandPersistenceAdapter(
    private val productCatalogCommandJpaRepository: ProductCatalogCommandJpaRepository,
    private val productCatalogQueryJooqRepository: ProductCatalogQueryJooqRepository,
    private val productItemCommandJpaRepository: ProductItemCommandJpaRepository,
    private val mapper: ProductCommandPersistenceMapper,
) : ProductCommandPort {
    override fun save(productCondition: SaveProductItemCondition): Product {
        val catalogId = productCondition.catalogId
        val productCatalogQueryResult =
            productCatalogQueryJooqRepository.findById(
                catalogId,
            ) ?: throw IllegalArgumentException("Product catalog not found with id: $catalogId")

        val productItemEntity = mapper.toEntity(productCondition, productCatalogQueryResult)

        val savedEntity = productItemCommandJpaRepository.save(productItemEntity)
        return mapper.toDomain(savedEntity)
    }

    override fun update(product: Product): Product {
        val catalogQueryResult =
            productCatalogQueryJooqRepository.findById(product.catalog.id) ?: throw IllegalArgumentException(
                "Product catalog not found with id: ${product.catalog.id}",
            )
        val entity = mapper.toEntity(product, catalogQueryResult)
        val savedEntity = productItemCommandJpaRepository.save(entity)
        return mapper.toDomain(savedEntity)
    }

    override fun deleteById(id: Long) {
        productItemCommandJpaRepository.deleteById(id)
    }
} 
